app/controllers/kaui/admin_tenants_controller.rb

```diff
@@ -415,6 +415,27 @@ def catalog_by_effective_date
     end
   end
 
+  def validate_plugin_name
+    json_response do
+      message = nil
+      entered_plugin_name = params.require(:plugin_name)
+      nodes_info = KillBillClient::Model::NodesInfo.nodes_info(options_for_klient) || []
+      plugins_info = nodes_info.first.plugins_info || []
+
+      found_plugin, weights = fuzzy_match(entered_plugin_name, plugins_info)
```

Comment:
I think this needs to somehow also take into account https://github.com/killbill/killbill-cloud/blob/master/kpm/lib/kpm/plugins_directory.yml.

For instance, if PayPal isn't installed and the user enters `killbill-paypal-` quickly (autocomplete doesn't have to show up), nothing is suggested.

Maybe an alternative design (that could help with the snowflakes issue above as well) would be to offer both a dropdown and a free text search? The dropdown would contain official OSS plugins as well as the ones already installed. The free text search would be for more advanced users with custom, proprietary, plugins.

Since 99% of the users would select from the list, maybe we could have some form of `toggle` that swaps from one to the other, making the dropdown the default?

Reply:
@pierre, I like the dropdown idea with a free text with suggestion.

For the dropdown:
* I was thinking creating a helper that will create a array of objects based on `https://github.com/killbill/killbill-cloud/blob/master/kpm/lib/kpm/plugins_directory.yml` using the artifact id to help the build of the `plugin name` and `plugin key` if it exist. 
* Plugin key because for paypal I notice that the key in the configuration is `paypal-express` and not ‘paypal’ .
* If no `artifact id` exist, add the `killbill-` into the index for the `plugin name` and the index as the `plugin key`.
* Using this array of object to populate the dropdown and when the upload button is clicked send all parameters including `configuration`, `plugin name`, `plugin type` and `plugin key`.
* Modify the endpoint to use the sent `plugin key` and `plugin name` as is, simplifying the method.

For the free text:
* Use the suggestion. Maybe the user misspell the name.
* Use the entered `plugin name` as the `plugin key`.